drivers/timer/npcx_itim_timer.c

```diff
@@ -0,0 +1,250 @@
+ *   "sleep/deep sleep" power state.
+ * - Provide a system clock timeout notification. In its ISR, the driver informs
+ *   the kernel that the specified number of ticks have elapsed.
+ * - Its prescaler is set to 1 and the formula between event timer's cycles and
+ *   ticks is 'cycles = (ticks * 32768) / CONFIG_SYS_CLOCK_TICKS_PER_SEC'
+ * - Compensate reading of ITIM64 which clock is gating after ec entered
+ *   "sleep/deep sleep" power state if CONFIG_PM is enabled.
+ */
+
+#include <drivers/clock_control.h>
+#include <drivers/timer/system_timer.h>
+#include <sys_clock.h>
+#include <spinlock.h>
+#include <soc.h>
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(itim, LOG_LEVEL_ERR);
+
+#define NPCX_ITIM32_MAX_CNT 0xffffffff
+#define NPCX_ITIM64_MAX_HALF_CNT 0xffffffff
+#define EVT_CYCLES_PER_SEC LFCLK /* 32768 Hz */
+#define SYS_CYCLES_PER_TICK (sys_clock_hw_cycles_per_sec() \
+					/ CONFIG_SYS_CLOCK_TICKS_PER_SEC)
+#define EVT_CYCLES_FROM_TICKS(ticks) \
+	ceiling_fraction(ticks * EVT_CYCLES_PER_SEC, \
+			 CONFIG_SYS_CLOCK_TICKS_PER_SEC)
+
+/* Instance of system and event timers */
+static struct itim64_reg *const sys_tmr = (struct itim64_reg *)
+					DT_INST_REG_ADDR_BY_NAME(0, sys_itim);
+static struct itim32_reg *const evt_tmr = (struct itim32_reg *)
+					DT_INST_REG_ADDR_BY_NAME(0, evt_itim);
+
+static const struct npcx_clk_cfg itim_clk_cfg[] = NPCX_DT_CLK_CFG_ITEMS_LIST(0);
+
+static struct k_spinlock lock;
+/* Announced cycles in system timer before executing z_clock_announce() */
+static uint64_t cyc_sys_announced;
+/* Current target cycles of time-out signal in event timer */
+static uint32_t cyc_evt_timeout;
+
+/* ITIM local inline functions */
+static inline uint64_t npcx_itim_get_sys_cyc64(void)
+{
+	uint32_t cnt64h, cnt64h_check, cnt64l;
+
+	/* Read 64-bit counter value from two 32-bit registers */
+	do {
+		cnt64h_check = sys_tmr->ITCNT64H;
+		cnt64l  = sys_tmr->ITCNT64L;
+		cnt64h = sys_tmr->ITCNT64H;
+	} while (cnt64h != cnt64h_check);
+
+	cnt64h = NPCX_ITIM64_MAX_HALF_CNT - cnt64h;
+	cnt64l = NPCX_ITIM64_MAX_HALF_CNT - cnt64l + 1;
+
+	/* Return current value of 64-bit counter value of system timer */
+	return (((uint64_t)cnt64h) << 32) | cnt64l;
+}
+
+/* ITIM local functions */
+static void npcx_itim_start_evt_tmr_by_tick(int32_t ticks)
+{
+	/*
+	 * Get desired cycles of event timer from the requested ticks which
+	 * round up to next tick boundary.
+	 */
+	if (ticks <= 0) {
+		ticks = 1;
+	}
+
+	if (ticks == K_TICKS_FOREVER) {
+		cyc_evt_timeout = NPCX_ITIM32_MAX_CNT;
+	} else {
+		cyc_evt_timeout = MIN(EVT_CYCLES_FROM_TICKS(ticks),
+				      NPCX_ITIM32_MAX_CNT);
+	}
+	LOG_DBG("ticks %x, cyc_evt_timeout %x", ticks, cyc_evt_timeout);
+
+	/* Disable event timer if needed and wait for it to take effect */
+	if (IS_BIT_SET(evt_tmr->ITCTS, NPCX_ITCTS_ITEN)) {
+		evt_tmr->ITCTS &= ~BIT(NPCX_ITCTS_ITEN);
+		while (IS_BIT_SET(evt_tmr->ITCTS, NPCX_ITCTS_ITEN))
+			;
```

Comment:
Can you put a upper bound on waiting of ITEN to set/clear for all instances?

Reply:
Done. Please see below. Thanks.